src/Std/Internal/Http/Data/Request.lean

```diff
@@ -0,0 +1,226 @@
+    toString req.version ++
+    "\r\n" ++
+    toString req.headers ++ "\r\n\r\n"
+
+/--
+Determines if a request method typically doesn't allow a request body
+-/
+@[inline]
+def isInformational (request : Head) : Prop :=
+  ¬request.method.allowsRequestBody
+
+/--
+Creates a new HTTP Request builder with default head (method: GET, version: HTTP/1.1, asterisk URI, empty headers)
+-/
+def new : Builder := { }
+
+namespace Builder
+
+/--
+Creates a new HTTP Request builder with default head (method: GET, version: HTTP/1.1, asterisk URI, empty headers)
+-/
+def empty : Builder := { }
+
+/--
+Sets the HTTP method for the request being built
+-/
+def method (builder : Builder) (method : Method) : Builder :=
+  { builder with head := { builder.head with method := method } }
+
+/--
+Sets the HTTP version for the request being built
+-/
+def version (builder : Builder) (version : Version) : Builder :=
+  { builder with head := { builder.head with version := version } }
+
+/--
+Sets the request target/URI for the request being built
+-/
+def uri (builder : Builder) (uri : RequestTarget) : Builder :=
+  { builder with head := { builder.head with uri := uri } }
+
+/--
+Sets the request target/URI for the request being built
+-/
+def uri! (builder : Builder) (uri : String) : Builder :=
+  { builder with head := { builder.head with uri := RequestTarget.parse! uri } }
+
+/--
+Adds a single header to the request being built
+-/
+def header (builder : Builder) (key : String) (value : String) : Builder :=
+  { builder with head := { builder.head with headers := builder.head.headers.insert key value } }
+
+/--
+Builds and returns the final HTTP Request with the specified body
+-/
+def body (builder : Builder) (body : t) : Request t :=
+  { head := builder.head, body := body }
+
+/--
+Builds and returns the final HTTP Request without a body
+-/
+def build (builder : Builder) : Request Body :=
+  { head := builder.head, body := .zero }
+
+/--
+Builds and returns the final HTTP Request with the specified body as binary data
+-/
+def binary (builder : Builder) (bytes : ByteArray) : Request Body :=
+  builder
+    |>.header "Content-Type" "application/octet-stream"
+    |>.body (Body.bytes bytes)
+
+/--
+Builds and returns the final HTTP Request with the specified body as plain text
+-/
+def text (builder : Builder) (body : String) : Request Body :=
+  builder
+    |>.header "Content-Type" "text/plain; charset=utf-8"
+    |>.body (body.toUTF8 |> Body.bytes)
+
+/--
+Builds and returns the final HTTP Request with the specified body as HTML
+-/
+def html (builder : Builder) (body : String) : Request Body :=
+  builder
+    |>.header "Content-Type" "text/html; charset=utf-8"
+    |>.body (body.toUTF8 |> Body.bytes)
+
+/--
+Builds and returns the final HTTP Request with form-encoded data
+-/
+def form (builder : Builder) (formData : List (String × String)) : Request Body :=
+  let encoded := formData.map (fun (k, v) => k ++ "=" ++ v) |>.foldl (· ++ "&" ++ ·) ""
```

Comment:
Surely this function needs to do some escaping of the data? Or is that supposed to happen elsewhere?

Reply:
It happens when it's converted to String.